examples/naive/aarch64/intt_dilithium_123_45678.s

```diff
@@ -0,0 +1,533 @@
+
+        // Interm. Reduction
+        montg_reduce data0
+        montg_reduce data1
+        montg_reduce data4
+        montg_reduce data5
+
+        // Layer 4
+        gs_butterfly data0, data4, root0, 0, 1
+        gs_butterfly data1, data5, root0, 0, 1
+        gs_butterfly data2, data6, root0, 0, 1
+        gs_butterfly data3, data7, root0, 0, 1
+
+        // Standard way using vector instructions
+
+        str_vi  data0, inp, (16*4)
+        str_vo  data1, inp, (-16*4 +  1*16)
+        str_vo  data2, inp, (-16*4 +  2*16)
+        str_vo  data3, inp, (-16*4 +  3*16)
+
+        str_vi  data4, inpp, (16*4)
+        str_vo  data5, inpp, (-16*4 +  1*16)
+        str_vo  data6, inpp, (-16*4 +  2*16)
+        str_vo  data7, inpp, (-16*4 +  3*16)
+
+        add inp, inp, #64
+        add inpp, inpp, #64
+
+        subs count, count, #1
+        cbnz count, layer45678_start
+
+// -----------------------------------------------------------------------------
+
+        ninv             .req v25
+        ninv_tw          .req v26
+        modulus_half     .req v30
+        neg_modulus_half .req v31
+
+        ASM_LOAD(xtmp, ninv_addr)
+        ld1r {ninv.4s}, [xtmp]
+        ASM_LOAD(xtmp, ninv_tw_addr)
+        ld1r {ninv_tw.4s}, [xtmp]
+
+        ushr modulus_half.4S, consts.4S, #1
+        neg neg_modulus_half.4S, modulus_half.4S
+
+        mov count, #8
+        ASM_LOAD(r_ptr0, roots_l012)
+        load_roots_123
+
+        .p2align 2
+layer123_start:
+
+        ldr_vo data0, in, 0
+        ldr_vo data1, in, (1*(1024/8))
+        ldr_vo data2, in, (2*(1024/8))
+        ldr_vo data3, in, (3*(1024/8))
+        ldr_vo data4, in, (4*(1024/8))
+        ldr_vo data5, in, (5*(1024/8))
+        ldr_vo data6, in, (6*(1024/8))
+        ldr_vo data7, in, (7*(1024/8))
+
+        gs_butterfly data0, data1, root1, 2, 3
+        gs_butterfly data2, data3, root2, 0, 1
+        gs_butterfly data4, data5, root2, 2, 3
+        gs_butterfly data6, data7, root3, 0, 1
+
+        gs_butterfly data0, data2, root0, 2, 3
+        gs_butterfly data1, data3, root0, 2, 3
+        gs_butterfly data4, data6, root1, 0, 1
+        gs_butterfly data5, data7, root1, 0, 1
+
+        // root0[0] includes ninv, manually computed.
+        gs_butterfly data0, data4, root0, 0, 1
+        gs_butterfly data1, data5, root0, 0, 1
+        gs_butterfly data2, data6, root0, 0, 1
+        gs_butterfly data3, data7, root0, 0, 1
+
+        canonical_reduce data4, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data5, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data6, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data7, modulus_half, neg_modulus_half, t2, t3
+
+        str_vo data4, in, (4*(1024/8))
+        str_vo data5, in, (5*(1024/8))
+        str_vo data6, in, (6*(1024/8))
+        str_vo data7, in, (7*(1024/8))        
+
+        mul_ninv data4, data5, data6, data7, data0, data1, data2, data3
```

Comment:
It would be interesting to double-check if the `mul_ninv` and the `canonical_reduce` can be replaced by a refined Barrett multiplication the sense of https://eprint.iacr.org/2022/439.pdf.